### PR TITLE
Implement MFA authentication APIs

### LINKED ISSUE #316: Support MFA APIs
### Checklist

- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

The current SDK contains only a portion of the Auth0 Authentication APIs. It's missing MFA APIs.

### Describe the ideal solution

Add support for MFA and the rest of APIs as described in the Auth0 Authentication APIs: https://auth0.com/docs/api/authentication

### Alternatives and current workarounds

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ewanharris**
### 🔧 Changes

Adds support for the `Challenge` and various `Verify*` APIs supported by the authentication API.

This does not add support for using the Authentication APIs as it was determined that the existing support on the management client is sufficient.

### 📚 References

Close #316

### 📝 Checklist

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### CODE REVIEW

File: authentication/authentication.go [hunk 1]
```diff
@@ -214,3 +224,127 @@ func (a *Authentication) UserInfo(ctx context.Context, accessToken string, opts
...
+		if err != nil {
+			return err
+		}
+
+		body.Set("client_assertion", clientAssertion)
+		body.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+		break
+	case params.ClientAssertion != "":
+		body.Set("client_assertion", params.ClientAssertion)
+		body.Set("client_assertion_type", params.ClientAssertionType)
```
Comment (Widcket) [inline authentication/authentication.go:272]: Should we check that `params.ClientAssertionType` is not empty? Like e.g. 

```go
case params.ClientAssertion != "" && params.ClientAssertionType != ""
```

File: authentication/mfa.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		return nil, err
+	}
+
+	err = m.authentication.Request(ctx, "POST", m.authentication.URI("oauth", "token"), data, &t, opts...)
+
+	return
+}
+
+// VerifyWithOOB verifies an MFA challenge using an out-of-band challenge (OOB), either Push notification,
+// SMS, or Voice.
```
Comment (Widcket) [inline authentication/mfa.go:75]:
```suggestion
// VerifyWithOOB verifies an MFA challenge using an out-of-band challenge (OOB), either push notification,
// SMS, or voice.
```

File: authentication/mfa/mfa.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// VerifyWithOTPRequest defines the request body for verifying an MFA challenge with OTP.
+type VerifyWithOTPRequest struct {
+	oauth.ClientAuthentication
+	MFAToken string
+	OTP      string
+}
+
+// VerifyWithOOBRequest defines the request body for verifying an MFA challenge with oob.
```
Comment (Widcket) [inline authentication/mfa/mfa.go:41]:
```suggestion
// VerifyWithOOBRequest defines the request body for verifying an MFA challenge with OOB.
```

File: authentication/mfa/mfa.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+type ChallengeRequest struct {
+	oauth.ClientAuthentication
+	// The token received from mfa_required error.
+	MFAToken string `json:"mfa_token,omitempty"`
+	// A whitespace-separated list of the challenges types accepted by your application.
+	// Accepted challenge types are oob or otp. Excluding this parameter means that your
+	// client application accepts all supported challenge type
+	ChallengeType string `json:"challenge_type,omitempty"`
+	// The ID of the authenticator to challenge. You can get the ID by querying the list of
+	// available authenticators for the user using `management.User.ListAuthenticationMethods`
```
Comment (Widcket) [inline authentication/mfa/mfa.go:17]:
```suggestion
	// The token received from the `mfa_required` error.
	MFAToken string `json:"mfa_token,omitempty"`
	// A whitespace-separated list of the challenges types accepted by your application.
	// Accepted challenge types are oob or otp. Excluding this parameter means that your
	// client application accepts all supported challenge type.
	ChallengeType string `json:"challenge_type,omitempty"`
	// The ID of the authenticator to challenge. You can get the ID by querying the list of
	// available authenticators for the user using `management.User.ListAuthenticationMethods`.
```

File: authentication/mfa/mfa.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	AuthenticatorID string `json:"authenticator_id,omitempty"`
+}
+
+// ChallengeResponse defines the response body when requesting an MFA Challenge.
```
Comment (Widcket) [inline authentication/mfa/mfa.go:21]:
```suggestion
// ChallengeResponse defines the response body when requesting an MFA challenge.
```

File: authentication/mfa/mfa.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
+package mfa
+
+import (
+	"github.com/auth0/go-auth0/authentication/oauth"
+)
+
+// ChallengeRequest defines the request body for requesting an MFA Challenge.
```
Comment (Widcket) [inline authentication/mfa/mfa.go:7]:
```suggestion
// ChallengeRequest defines the request body for requesting an MFA challenge.
```

File: authentication/mfa/mfa.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+// ChallengeResponse defines the response body when requesting an MFA Challenge.
+type ChallengeResponse struct {
+	// The type of challenge requested.
+	ChallengeType string `json:"challenge_type,omitempty"`
+	// The OOB code to use when calling `VerifyWithOOBRequest`
+	// Only present when ChallengeType is "oob".
+	OOBCode string `json:"oob_code,omitempty"`
+	/// If included, then the user should be prompted for a `BindingCode` which should be included
+	// in the VerifyWithOOBRequest provided to VerifyWithOOB.
+	// Only present when ChallengeType is "oob".
```
Comment (Widcket) [inline authentication/mfa/mfa.go:30]:
```suggestion
	// Only present when `ChallengeType` is "oob".
	OOBCode string `json:"oob_code,omitempty"`
	/// If included, then the user should be prompted for a `BindingCode` which should be included
	// in the `VerifyWithOOBRequest` provided to `VerifyWithOOB`.
	// Only present when `ChallengeType` is "oob".
```

File: authentication/authentication.go (hunk 1 above)
Comment (ewanharris) [reply authentication/authentication.go:272]: Yeah that makes sense, will fix

Comment (ewanharris) [reply authentication/authentication.go:272]: Added in 716bdadcfdeeafb9a3bd86d11f384a8d7937a59e and also took the opportunity to extend `addClientAuthenticationToClientAuthStruct` to allow specifying if a client secret/client assertion is required

**Review by Widcket — APPROVED** [review]

**Pull request merged**